gtsfm/two_view_estimator.py

```diff
@@ -360,6 +371,67 @@ def run_2view(
         return post_isp_i2Ri1, post_isp_i2Ui1, post_isp_v_corr_idxs, pre_ba_report, post_ba_report, post_isp_report
 
 
+def optimize_essential_matrix(
+    keypoints_i1: Keypoints,
+    keypoints_i2: Keypoints,
+    corr_idxs_i1i2: np.ndarray,
+    camera_intrinsics_i1: gtsam.Cal3Bundler,
+    camera_intrinsics_i2: gtsam.Cal3Bundler,
+    i2Ri1: Rot3,
+    i2Ui1: Unit3,
+    use_robust_loss: bool = True,
+) -> Tuple[Dict[Tuple[int, int], Rot3], Dict[Tuple[int, int], Unit3]]:
+    """Jointly optimize over all essential matrices for edges contained in triplets.
+
+    Args:
+        triplets: List of triplets in view graph.
+        i2Ri1_dict: Dict from (i1, i2) to relative rotation of i1 with respect to i2.
+        i2Ui1_dict: Dict from (i1, i2) to relative translation direction of i1 with respect to i2.
+        calibrations: List of calibrations for each image.
+        corr_idxs_i1i2: Dict from (i1, i2) to indices of verified correspondences from i1 to i2.
+        keypoints: keypoints for each images.
+        use_robust_loss: whether to use Huber loss.
+
+    Returns:
+        Optimized relative rotations and directions.
+    """
+    # Create GTSAM objects.
+    graph = gtsam.NonlinearFactorGraph()
+    initial = gtsam.Values()
+    noise_model=gtsam.noiseModel.Isotropic.Sigma(1, 1e-2)
+    if use_robust_loss:
+        huber_loss = gtsam.noiseModel.mEstimator.Huber.Create(1.345)
+        noise_model = gtsam.noiseModel.Robust.Create(huber_loss, noise_model)
+
+    # Add essential matrix factors.
+    mkps1 = keypoints_i1.coordinates[corr_idxs_i1i2[:, 0]]
+    mkps2 = keypoints_i2.coordinates[corr_idxs_i1i2[:, 1]]
+    initial.insert(0, gtsam.EssentialMatrix(i2Ri1, i2Ui1))
+    for kp1, kp2 in zip(mkps1, mkps2):
+        graph.add(
+            gtsam.EssentialMatrixFactor(
```

Comment:
we should probably use the newer implementations @dellaert added. I think some bugs with the essential matrix were fixed too?